dali/operators/segmentation/random_object_bbox.cc

```diff
@@ -0,0 +1,442 @@
+  .AddOptionalArg<vector<float>>("class_weights", R"(Relative probabilities of foreground classes.
+
+Each value corresponds to a class label in ``classes`` or a 1-based number if ``classes`` are
+not specified.
+The values are normalized so that they sum to 1.)", nullptr, true)
+  .AddOptionalArg<int>("k_largest", "If specified, at most k largest bounding boxes are consider",
+    nullptr)
+  .AddOptionalArg<vector<int>>("threshold", R"(Minimum extent(s) of the bounding boxes to return.
+
+If current class doesn't contain any bounding box that meets this condition, the largest one
+is returned.)", nullptr, true)
+  .AddOptionalArg("format", R"(Format in which the data is returned.
+
+Possible choices are::
+  * "anchor_shape" (the default) - there are two outputs: anchor and shape
+  * "start_end" - there are two outputs - bounding box start and one-past-end coordinates
+  * "box" - ther'es one output that contains concatenated start and end coordinates
+)", "anchor_shape");
+
+
+bool RandomObjectBBox::SetupImpl(vector<OutputDesc> &out_descs, const HostWorkspace &ws) {
+  out_descs.resize(spec_.NumOutput());
+  auto in_shape = ws.InputRef<CPUBackend>(0).shape();
+  int ndim = in_shape.sample_dim();
+  int N = in_shape.num_samples();
+
+  DALI_ENFORCE(N == 0 || (ndim >= 1 && ndim <= 6),
+      make_string("Unsuported number of dimensions ", ndim, "; must be 1..6"));
+  AcquireArgs(ws, N, ndim);
+  out_descs[0].type = TypeTable::GetTypeInfo(DALI_INT32);
+  out_descs[0].shape = uniform_list_shape<DynamicDimensions>(
+      N, TensorShape<1>{ format_ == Out_Box ? 2*ndim : ndim });
+  if (format_ != Out_Box)
+    out_descs[1] = out_descs[0];
+  if (HasClassLabelOutput()) {
+    out_descs[class_output_idx_].type = TypeTable::GetTypeInfo(DALI_INT32);
+    out_descs[class_output_idx_].shape.resize(N, 0);
+  }
+
+  return true;
+}
+
+void RandomObjectBBox::AcquireArgs(const HostWorkspace &ws, int N, int ndim) {
+  background_.Acquire(spec_, ws, N);
+  if (classes_.IsDefined())
+    classes_.Acquire(spec_, ws, N);
+  foreground_prob_.Acquire(spec_, ws, N);
+  if (weights_.IsDefined())
+    weights_.Acquire(spec_, ws, N);
+  if (threshold_.IsDefined())
+    threshold_.Acquire(spec_, ws, N, TensorShape<1>{ndim});
+
+  if (weights_.IsDefined() && classes_.IsDefined()) {
+    DALI_ENFORCE(weights_.get().shape == classes_.get().shape, make_string(
+      "If both ``classes`` and ``class_weights`` are provided, their shapes must match. Got:"
+      "\n  classes.shape  = ", classes_.get().shape,
+      "\n  weights.shape  = ", weights_.get().shape));
+  }
+}
+
+
+#define INPUT_TYPES (bool, uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t)
+
+template <typename T>
+void RandomObjectBBox::FindLabels(std::unordered_set<int> &labels, const T *data, int64_t N) {
+  if (!N)
+    return;
+  T prev = data[0];
+  labels.insert(prev);
+  for (int64_t i = 1; i < N; i++) {
+    if (data[i] == prev)
```

Comment:
Well... constant - yes; short - not necessarily. It will certainly take longer than a simple comparison with previous value. Given the nature of the data (segmentation masks), we're very likely to see long runs of repeated values, so optimizing for that certainly has merit.